torch/distributed/_cuda_p2p/__init__.py

```diff
@@ -0,0 +1,123 @@
+
+    # Using ProcessGroupCudaP2P while specifying ProcessGroupCudaP2P.Options
+    pg_options = ProcessGroupCudaP2P.Options()
+    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)
+
+    # Using ProcessGroupCudaP2P while specifying ProcessGroupNCCL.Options
+    pg_options = ProcessGroupNCCL.Options()
+    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)
+
+    # Using ProcessGroupCudaP2P while specifying both
+    # ProcessGroupCudaP2P.Options and ProcessGroupNCCL.Options
+    pg_options = ProcessGroupCudaP2P.Options()
+    pg_options.nccl_options = ProcessGroupNCCL.Options()
+    dist.init_process_group(backend="cuda_p2p", pg_options=pg_options, ...)
+
+    # Down-casting the backend to access p2p buffers for cuda_p2p specific
+    # optimizations
+    if is_cuda_p2p_group(group):
+        backend = get_cuda_p2p_backend(group)
+        if required_p2p_buffer_size > backend.get_buffer_size():
+            # fallback
+        p2p_buffer = backend.get_p2p_buffer(...)
+    else:
+        # fallback
+"""
+
+
+def _create_cuda_p2p_group(
+    dist_backend_opts: "_DistributedBackendOptions",
+    options: Union[
+        "c10d.ProcessGroupCudaP2P.Options", "c10d.ProcessGroupNCCL.Options", None
+    ],
+) -> "Backend":
+    if not c10d.is_nccl_available():
+        raise RuntimeError("The cuda_p2p backend is not available")
+    if options is None:
+        options = c10d.ProcessGroupCudaP2P.Options()
+        options.nccl_options = c10d.ProcessGroupNCCL.Options()
+    elif isinstance(options, c10d.ProcessGroupNCCL.Options):
+        nccl_options = options
+        options = c10d.ProcessGroupCudaP2P.Options()
+        options.nccl_options = nccl_options
+    elif isinstance(options, c10d.ProcessGroupCudaP2P.Options):
+        if options.nccl_options is None:
+            options.nccl_options = c10d.ProcessGroupNCCL.Options()
+    else:
+        raise TypeError(
+            "options for cuda_p2p must be ProcessGroupCudaP2P.Options "
+            f"or ProcessGroupNCCL.Options (got: {type(options)})"
+        )
+
+    return c10d.ProcessGroupCudaP2P(
+        dist_backend_opts.store,
+        dist_backend_opts.group_rank,
+        dist_backend_opts.group_size,
+        options,
+    )
+
+
+def is_cuda_p2p_group(group: c10d.ProcessGroup) -> bool:
+    if not c10d.is_nccl_available():
+        return False
+    try:
+        backend = group._get_backend(torch.device("cuda"))
+    except Exception:
+        return False
+    return isinstance(backend, c10d.ProcessGroupCudaP2P) and backend.is_p2p_available()
+
+
+def get_cuda_p2p_backend(group: c10d.ProcessGroup) -> "c10d.ProcessGroupCudaP2P":
+    if not is_cuda_p2p_group(group):
+        raise TypeError("group is not a cuda_p2p process group.")
+    return cast(
+        c10d.ProcessGroupCudaP2P,
+        group._get_backend(torch.device("cuda")),
+    )
+
+
+def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
```

Comment:
Wondering why we need this method as I see no callsites to it?

Reply:
Hey @wanchaol, it's used in the subsequent PR. But I should add a test for it in this PR.